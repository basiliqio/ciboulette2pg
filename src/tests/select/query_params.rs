use super::*;

#[ciboulette2postgres_test]
async fn sparse(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    let data = init_values::init_values(&mut transaction).await;
    let people_id = data.get("peoples").unwrap().first().unwrap();
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples/{}?fields[peoples]=first-name", people_id).as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn include(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    let data = init_values::init_values(&mut transaction).await;
    let people_id = data.get("peoples").unwrap().first().unwrap();
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples/{}?include=articles", people_id).as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn include_full_path(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    let data = init_values::init_values(&mut transaction).await;
    let people_id = data.get("peoples").unwrap().first().unwrap();
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples/{}?include=peoples.articles", people_id).as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn include_multiple_resources(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    let data = init_values::init_values(&mut transaction).await;
    let people_id = data.get("peoples").unwrap().first().unwrap();
    let raw_rows = test_select(
        &mut transaction,
        format!(
            "/peoples/{}?include=peoples.articles,people-article",
            people_id
        )
        .as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn include_multiple_resources_with_sparsing(
    mut transaction: sqlx::Transaction<'_, sqlx::Postgres>
) {
    let data = init_values::init_values(&mut transaction).await;
    let people_id = data.get("peoples").unwrap().first().unwrap();
    let raw_rows = test_select(&mut transaction, format!("/peoples/{}?include=peoples.articles,people-article&fields[peoples]=last-name&fields[articles]=title&fields[people-article]=article_id,people_id", people_id).as_str()).await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn sorting(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    init_values::init_values(&mut transaction).await;
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples?&sort=first-name").as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn sorting_full_path(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    init_values::init_values(&mut transaction).await;
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples?sort=peoples.first-name").as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn sorting_desc(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    init_values::init_values(&mut transaction).await;
    let raw_rows = test_select(&mut transaction, format!("/peoples?sort=-age").as_str()).await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn sorting_multiple_fields(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    init_values::init_values(&mut transaction).await;
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples?sort=age,-first-name").as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    check_rows(&res);
}

#[ciboulette2postgres_test]
async fn sorting_multiple_resources(mut transaction: sqlx::Transaction<'_, sqlx::Postgres>) {
    init_values::init_values(&mut transaction).await;
    let raw_rows = test_select(
        &mut transaction,
        format!("/peoples?include=articles&sort=articles.title,first-name").as_str(),
    )
    .await;
    let res =
        Ciboulette2PostgresRow::from_raw(&raw_rows).expect("to deserialize the returned rows");
    let res = res
        .into_iter()
        .filter(|x| x.type_() == &"peoples")
        .collect(); // FIXME don't
    check_rows(&res);
}
