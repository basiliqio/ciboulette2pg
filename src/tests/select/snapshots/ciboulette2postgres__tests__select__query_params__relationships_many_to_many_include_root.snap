---
source: src/tests/select/query_params.rs
expression: value

---
[
  {
    "id": "[uuid]",
    "type": "articles",
    "related_type": "peoples",
    "related_id": "[uuid]"
  },
  {
    "id": "[uuid]",
    "type": "articles",
    "related_type": "peoples",
    "related_id": "[uuid]"
  },
  {
    "id": "[uuid]",
    "type": "peoples"
  },
  {
    "id": "[uuid]",
    "type": "author",
    "data": {
      "age": 22,
      "first-name": "Francis",
      "gender": "M",
      "last-name": "Le Roy",
      "twitter": null
    },
    "related_type": "articles",
    "related_id": "[uuid]"
  },
  {
    "id": "[uuid]",
    "type": "author",
    "data": {
      "age": 22,
      "first-name": "Francis",
      "gender": "M",
      "last-name": "Le Roy",
      "twitter": null
    },
    "related_type": "articles",
    "related_id": "[uuid]"
  },
  {
    "id": "[uuid]",
    "type": "author",
    "data": {
      "age": 34,
      "first-name": "Somebody",
      "gender": "F",
      "last-name": "Wuhu",
      "twitter": "@randomhandle"
    },
    "related_type": "articles",
    "related_id": "[uuid]"
  }
]
